contracts/Shortfall/Shortfall.sol

```diff
@@ -71,34 +100,133 @@ contract Shortfall is OwnableUpgradeable, ReentrancyGuardUpgradeable {
+        require(
+            block.number > auction.highestBidBlock + nextBidderBlockLimit && auction.highestBidder != address(0),
+            "waiting for next bidder. cannot close auction"
+        );
+
+        uint256 marketsCount = auction.markets.length;
+        uint256[] memory marketsDebt = new uint256[](marketsCount);
+
+        auction.status = AuctionStatus.ENDED;
+
+        for (uint256 i; i < marketsCount; ++i) {
+            VToken vToken = VToken(address(auction.markets[i]));
+            IERC20Upgradeable erc20 = IERC20Upgradeable(address(vToken.underlying()));
+
+            if (auction.auctionType == AuctionType.LARGE_POOL_DEBT) {
+                uint256 bidAmount = ((auction.marketDebt[auction.markets[i]] * auction.highestBidBps) / MAX_BPS);
+                erc20.safeTransfer(address(auction.markets[i]), bidAmount);
+                marketsDebt[i] = bidAmount;
+            } else {
+                erc20.safeTransfer(address(auction.markets[i]), auction.marketDebt[auction.markets[i]]);
+                marketsDebt[i] = auction.marketDebt[auction.markets[i]];
+            }
+
+            auction.markets[i].badDebtRecovered(auction.marketDebt[auction.markets[i]]);
+        }
+
+        uint256 riskFundBidAmount = auction.seizedRiskFund;
+
+        if (auction.auctionType == AuctionType.LARGE_POOL_DEBT) {
+            riskFund.transferReserveForAuction(comptroller, riskFundBidAmount);
+            BUSD.safeTransfer(auction.highestBidder, riskFundBidAmount);
+        } else {
+            riskFundBidAmount = (auction.seizedRiskFund * auction.highestBidBps) / MAX_BPS;
+            uint256 remainingRiskFundSeizedAmount = auction.seizedRiskFund - riskFundBidAmount;
+            riskFund.transferReserveForAuction(comptroller, auction.seizedRiskFund - remainingRiskFundSeizedAmount);
+            BUSD.safeTransfer(auction.highestBidder, auction.seizedRiskFund - remainingRiskFundSeizedAmount);
+        }
+
+        emit AuctionClosed(
+            comptroller,
+            auction.highestBidder,
+            auction.highestBidBps,
+            riskFundBidAmount,
+            auction.markets,
+            marketsDebt
+        );
+    }
+
+    /**
+     * @notice Restart an auction
+     * @param comptroller Address of the pool
+     * @custom:events Emits AuctionRestarted event on successful restart
+     */
+    function restartAuction(address comptroller) external {
+        Auction storage auction = auctions[comptroller];
+
+        require(auction.startBlock != 0 && auction.status == AuctionStatus.STARTED, "no on-going auction");
+        require(
+            block.number > auction.startBlock + waitForFirstBidder && auction.highestBidder == address(0),
+            "you need to wait for more time for first bidder"
+        );
+
+        auction.status = AuctionStatus.ENDED;
+
+        emit AuctionRestarted(comptroller);
+        startAuction(comptroller);
+    }
```

Comment:
Let's move `initialize` closer to the constructor. It's a common practice since the initializer function is sort-of constructor for upgradeable contracts.

Reply:
I agree with. It does break the sort order

Reply:
Let's make it `external`? 🤷